SpatialGDK/Source/SpatialGDKServices/Private/LocalReceptionistProxyServerManager.cpp

```diff
@@ -0,0 +1,217 @@
+	
+	return false;
+}
+
+bool FLocalReceptionistProxyServerManager::CheckIfPortIsBound(int32 Port, FString& PID, FText& LogMsg)
+{
+	const FString NetStatCmd = TEXT("netstat");
+
+	// -a display active tcp/udp connections, -o include PID for each connection, -n don't resolve hostnames
+	const FString NetStatArgs = TEXT("-n -o -a");
+	FString NetStatResult;
+	int32 ExitCode;
+	FString StdErr;
+	FString ProcessName;
+	bool bSuccess = FPlatformProcess::ExecProcess(*NetStatCmd, *NetStatArgs, &ExitCode, &NetStatResult, &StdErr);
+
+	if (ExitCode == ExitCodeSuccess && bSuccess)
+	{
+		// Get the line of the netstat output that contains the port we're looking for.
+		FRegexPattern PidMatcherPattern(FString::Printf(TEXT("(.*?:%i.)(.*)( [0-9]+)"), Port));
+		FRegexMatcher PidMatcher(PidMatcherPattern, NetStatResult);
+		if (PidMatcher.FindNext())
+		{
+			const FString State = PidMatcher.GetCaptureGroup(2 /* Get the State of the process, which is the second group. */);
+			if (State.Contains("LISTENING"))
+			{
+				
+				PID = PidMatcher.GetCaptureGroup(3 /* Get the PID, which is the third group. */);
+				if (GetProcessName(PID, ProcessName))
+				{
+					LogMsg = FText::Format(LOCTEXT("ProcessBlockingPort", "{0} process with PID:{1}."), FText::FromString(ProcessName), FText::FromString(PID));
+					return true;
+				}
+
+				LogMsg = FText::Format(LOCTEXT("ProcessBlockingPort", "Unknown process with PID:{1}."), FText::FromString(PID));
+			
+				return true;
+			}
+		}
+	}
+	else
+	{
+		LogMsg = FText::Format(LOCTEXT("ProcessBlockingPort", "Failed to check if any process is blocking required port. Error:{0}"), FText::FromString(StdErr));
+	}
+
+	LogMsg = LOCTEXT("NoProcessBlockingProxyPort", "No Process is blocking the required port.");
+
+	return false;
+}
+
+
+bool FLocalReceptionistProxyServerManager::TryKillBlockingPortProcess(const FString& PID)
+{
+	bool bSuccess = false;
+
+	int32 ExitCode;
+	FString StdErr;
+
+	const FString TaskKillCmd = TEXT("taskkill");
+	const FString TaskKillArgs = FString::Printf(TEXT("/F /PID %s"), *PID);
+	FString TaskKillResult;
+	bSuccess = FPlatformProcess::ExecProcess(*TaskKillCmd, *TaskKillArgs, &ExitCode, &TaskKillResult, &StdErr);
+	bSuccess = bSuccess && ExitCode == ExitCodeSuccess;
+	if (!bSuccess)
+	{
+		const FText ErrorMessage = FText::Format(LOCTEXT("FailedToKillBlockingProcess", "Failed to kill process blocking required port. Error: '{0}'"), FText::FromString(StdErr));
+		UE_LOG(LogLocalReceptionistProxyServerManager, Error, TEXT("%s"),*ErrorMessage.ToString());
+	}
+
+	return bSuccess;
+}
+
+bool FLocalReceptionistProxyServerManager::LocalReceptionistProxyServerPreRunChecks(int32 ReceptionistPort)
+{
+	FText OutLogMessage;
+	FString PID;
+
+	//Check if any process is blocking the receptionist port
+	if(CheckIfPortIsBound(ReceptionistPort, PID, OutLogMessage))
+	{
+		//Try killing the process that blocks the receptionist port 
+		bool bProcessKilled = TryKillBlockingPortProcess(PID);
```

Comment:
We should ask for user confirmation before killing the process.

Reply:
I am calling this when we start up the EditorToolbarModule so I do not think that getting user confirmation is possible at this stage? Do you think I should move it somewhere else?

Reply:
Hmm why don't we do this right before starting the proxy?

Reply:
In the case that the engine crashes when the proxy is running, then the proxy process will stay running on the background. Then when you open the editor the Local deployment will fail to run. So to solve that issue every time we start the editor we make sure we kill the process.